src/codegen/mod.rs

```diff
@@ -798,6 +796,105 @@ impl<'a> Bitfield<'a> {
             methods.extend(items.into_iter());
             offset += width;
         }
+
+        bitfield_layout
+    }
+}
+
+struct StructLayout<'a, 'ctx: 'a> {
+    ctx: &'a BindgenContext<'ctx>,
+    comp: &'a CompInfo, 
+    latest_offset: usize,
+    padding_count: usize,
+}
+
+impl<'a, 'ctx> StructLayout<'a, 'ctx> {
+    pub fn new(ctx: &'a BindgenContext<'ctx>, comp: &'a CompInfo) -> Self {
+        StructLayout {
+            ctx: ctx,
+            comp: comp,
+            latest_offset: 0,
+            padding_count: 0,
+        }
+    }
+
+    pub fn saw_vtable(&mut self, _vtable: &Vtable) {
+        self.latest_offset += mem::size_of::<* mut ()>();
+    }
+
+    pub fn saw_base(&mut self, base_ty: &Type) {
+        if let Some(base_layout) = base_ty.layout(self.ctx) {
+            self.latest_offset += self.padding_bytes(base_layout) +
+                                    base_ty.calc_size(self.ctx).unwrap_or(base_layout.size);
```

Comment:
Yes, we could trust `base_layout.size` here. In fact, I use `base_ty.calc_size` because I have tried to fix those virtual inheritance vtable padding issues, but I remove those code before submit, because it is a little complex, hard to cover everything in one patch :(